### PR TITLE
[BUGFIX] TSDB: Fix query overlapping in-order and ooo head 

### PR DESCRIPTION
**Pull request by bboreham**
This is competing with #14681, and includes the test added there by @krajorama 

The basic idea is that `getOOOSeriesChunks` can populate `Meta.Chunk`, but since it only returns one `Meta` per overlapping time-slot, that pointer may end up in a `Meta` with a head-chunk ID. So we need `HeadAndOOOChunkReader.ChunkOrIterable()` to call `mergedChunks` in that case.

Previously, `mergedChunks` was checking that `meta.Ref` was a valid OOO chunk reference, but it never actually uses that reference; it just finds all chunks overlapping in time.  So we can delete that code.

### CODE REVIEW

**Review by fionaliao — DISMISSED** [review]

File: tsdb/head_read.go [hunk 1]
```diff
@@ -481,31 +481,12 @@ func (s *memSeries) chunk(id chunks.HeadChunkID, chunkDiskMapper *chunks.ChunkDi
...
 // This function is not thread safe unless the caller holds a lock.
 // The caller must ensure that s.ooo is not nil.
 func (s *memSeries) mergedChunks(meta chunks.Meta, cdm *chunks.ChunkDiskMapper, hr *headChunkReader, mint, maxt int64, maxMmapRef chunks.ChunkDiskMapperRef) (chunkenc.Iterable, error) {
-	_, cid, _ := unpackHeadChunkRef(meta.Ref)
-
-	// ix represents the index of chunk in the s.mmappedChunks slice. The chunk meta's are
-	// incremented by 1 when new chunk is created, hence (meta - firstChunkID) gives the slice index.
-	// The max index for the s.mmappedChunks slice can be len(s.mmappedChunks)-1, hence if the ix
-	// is len(s.mmappedChunks), it represents the next chunk, which is the head chunk.
-	ix := int(cid) - int(s.ooo.firstOOOChunkID)
```
Comment (krajorama) [inline tsdb/head_read.go:498]: I don't see from git history why this check was here, maybe a bug during development, we'd need to ask @jesusvazquez or @codesome . If we wanted to keep it, we could use the third return value from unpackHeadChunkRef, isOOO to decide whether to run the check or not.

File: tsdb/ooo_head_read.go [hunk 2]
```diff
@@ -253,6 +253,10 @@ func (cr *HeadAndOOOChunkReader) ChunkOrIterable(meta chunks.Meta) (chunkenc.Chu
 	}
 
 	s.Lock()
+	if !isOOO && meta.Chunk == nil && s.ooo == nil { // Don't need to merge if no OOO chunks.
```
Comment (krajorama) [inline tsdb/ooo_head_read.go:256]: I'm trying to understand the performance implication of `s.ooo == nil`. It seems reasonable to have that condition, but it means that we'll try the merge more often than before. Does it make it better/same/worse than before #14396 ?

**Review by krajorama — APPROVED** [review]
LGTM

File: tsdb/ooo_head_read.go (hunk 2 above)
Comment (bboreham) [reply tsdb/ooo_head_read.go]: I don't think #14396 changed performance much, and didn't change merging at all.
The big difference in performance and the way merges happen is in #14354.

**Pull request merged**